KYC Service Layer

kaapstorm opened the pull request:
## Technical Summary

Jira: [SC-4128](https://dimagi.atlassian.net/browse/SC-4128), [SC-4139](https://dimagi.atlassian.net/browse/SC-4139)
Base branch: https://github.com/dimagi/commcare-hq/pull/35765

This change adds a service layer to the KYC integration that makes API calls and parses the responses.

## Feature Flag

KYC_VERIFICATION

## Safety Assurance

### Safety story

We are unable to test API calls yet. We are waiting for a service provider to grant us access to the API.

### Automated test coverage

Tests included. More tests to follow once we have settled on return values.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change


[SC-4128]: https://dimagi.atlassian.net/browse/SC-4128?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[SC-4139]: https://dimagi.atlassian.net/browse/SC-4139?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kaapstorm committed:
Add `KycConfig.get_connection_settings()`

kaapstorm committed:
`_row_data()` -> `KycConfig.get_user_objects()`

kaapstorm committed:
Move `clean()` from form to model

Form validation still works because it is a ModelForm.

kaapstorm committed:
API call to verify user

kaapstorm committed:
Move tests

kaapstorm committed:
`KycConfig.get_user_objects_by_ids()`

kaapstorm commented:
To avoid confusion, please don't merge until https://github.com/dimagi/commcare-hq/pull/35765 has been merged.

ajeety4 commented on settings.py:
```diff
@@ -1172,6 +1172,14 @@ def _pkce_required(client_id):
 PERMANENT_DELETION_WINDOW = 30  # days
 
+# Used by `corehq.apps.integration.kyc`. Override in localsettings.py
+MTN_KYC_CONNECTION_SETTINGS = {
```
(I am probably behind on recent updates with MTN )
What is the rationale for having a common connection settings ? With this not be project specific ?

kaapstorm replied:
It turns out that the API service provider will have a contract with the platform provider (i.e. Dimagi), and will grant access to them, and not to each project ...

... which has just made me think that saving the ConnectionSettings instance is not a good idea. Projects/users can't read the secrets of ConnectionSettings instances, but they can change them. So either we need to have a way to write-protect ConnectioinSettings, or we need to create them on the fly without persisting them.

ajeety4 commented on corehq/apps/integration/kyc/tests/test_models.py:
```diff
@@ -47,3 +52,76 @@ def test_bad_config(self):
+            'other_case_property': 'other_case_value',
+        }
+
+    def test_assert_other_case_type(self):
```
leftover ?

kaapstorm replied:
Just verifies behavior. But if you think I should drop that behavior, then I'll drop this test.

ajeety4 replied:
Ah, I  see this was a placeholder (with just a `pass` statement) and then was updated in the later commit. 

Good with the test.

ajeety4 commented on corehq/apps/integration/kyc/models.py:
```diff
@@ -52,6 +53,21 @@ class Meta:
             models.UniqueConstraint(fields=['domain', 'provider'], name='unique_domain_provider'),
         ]
 
+    def clean(self):
```
Cool :+1:

ajeety4 commented on corehq/apps/integration/kyc/models.py:
```diff
@@ -82,7 +98,7 @@ def get_user_objects(self):
             return CommCareUser.by_domain(self.domain)
         elif self.user_data_store == UserDataStore.OTHER_CASE_TYPE:
-            # assert self.other_case_type
+            assert self.other_case_type
```
wondering if this is required since we have validation now at model level

kaapstorm replied:
I put this in because `Model.clean()` gets called when a form is saved, but it isn't called by `Model.save()`. So it's possible for `self.other_case_type` not to have a value if a dev forgets to set it in a test, or a script, etc. It's just a way to indicate to a dev why this would break.

ajeety4 replied:
> I put this in because Model.clean() gets called when a form is saved, but it isn't called by Model.save()
Ah noted. I thought it was called on save as well.

ajeety4 commented on corehq/apps/integration/kyc/services.py:
```diff
@@ -1,10 +1,141 @@
+def verify_users(user_objs, config):
+    # TODO: An endpoint to verify a group of users does not seem to be
+    #       available using Chenosis. If we have to do this with
+    #       multiple calls, consider using Celery gevent workers.
```
:100:

ajeety4 commented on corehq/apps/integration/kyc/services.py:
```diff
@@ -1,10 +1,141 @@
+    for user_obj in user_objs:
+        is_verified = verify_user(user_obj, config)
+        save_result(user_obj, config, is_verified)
+        results.append(is_verified)
```
We should also add an identifier (e.g. `id`) along with the verification result. 

```
results.append(
    {
        'id': 'abc',
        'is_verified': True,
    }
)
```

ajeety4 replied:
Marking this as resolved as this was implemented in the [PR for metrics](https://github.com/dimagi/commcare-hq/pull/35853) where this was needed.

ajeety4 commented on corehq/apps/integration/kyc/services.py:
```diff
@@ -1,10 +1,141 @@
+            config.domain,
+            case.case_id,
+            case_properties=update,
+            device_id='corehq.apps.integration.kyc.services.save_result',
```
nit:  Consider using  `device_id=__name__ + '.save_result'`

ajeety4 replied:
This is not needed as some refactoring was done that uses a new KycUser Class in the following [PR for KYC User metrics](https://github.com/dimagi/commcare-hq/pull/35853)

ajeety4 reviewed (commented):
Thanks Norman for working on this. 
This is looking quite good. Left some minor queries and suggestions.

Charl1996 commented on corehq/apps/integration/kyc/services.py:
```diff
@@ -1,10 +1,129 @@
+
+    # TODO: Determine what thresholds we want
+    required_thresholds = {
+        'firstName': 100,
```
@kaapstorm How does these scores function?

kaapstorm replied:
It looks like a percentage of how close the value that was submitted is to the value that MTN has on record. We haven't been able to test this yet, because we don't (yet) have credentials we can use for testing.

Charl1996 commented:
Just a note:

I see a lot of places where we check the `user_data_store` on the `KycProviders` config. This makes me think that something like the [adapter pattern](https://refactoring.guru/design-patterns/adapter) might come in handy here where the service makes use of the adapter to interact with the various underlying data store models. That way we could always just query the adapter for data-store-specific actions/data instead of having if-else blocks. The adapter pattern will also make is very easy to add new data stores, since we would only have to implement the new adapter with the various ways in which data can be read from/written to this potentially new data store without having to update if-else blocks.

That said, I think the approach in this PR is OK as is since I don't think this particular feature will see new data stores being added frequently (if ever).

kaapstorm committed:
Default to "custom", fall back to "standard"

If a field is in a user's usercase, or custom user data, then use that,
otherwise fall back to a CommCareUser property. Only allow safe
properties.

kaapstorm committed:
Drop "source"

kaapstorm committed:
Merge pull request #35804 from dimagi/nh/mapping

Safe mapping

ajeety4 commented on corehq/apps/integration/kyc/services.py:
```diff
@@ -1,36 +1,174 @@
+
+def save_result(user_obj, config, is_verified):
+    update = {
+        'kyc_provider': config.provider.value,  # TODO: Or config.provider.label?
```
`config.provider.value` does not work as provider is the actual string value. 

```suggestion
        'kyc_provider': config.provider,  # TODO: Or config.provider.get_provider_display()?
```

ajeety4 replied:
I am bit inclined towards storing the value here in case we want to make a conditional check on this later.

kaapstorm commented:
> the [adapter pattern](https://refactoring.guru/design-patterns/adapter) might come in handy here

I think this is a good idea. It might be out of scope for this PR, but worth looking at in a follow-up PR.

kaapstorm committed:
`config.provider` is the string value

kaapstorm committed:
Don't save system ConnectionSettings

ajeety4 committed:
fix: datetime string for json field

ajeety4 committed:
minor fix: save() on  user data

ajeety4 commented:
Hey @kaapstorm , I hope you don't mind. I added a minor fix (https://github.com/dimagi/commcare-hq/pull/35774/commits/16583f1e99a99d0abb4202c945fc83efaf8060f5 and https://github.com/dimagi/commcare-hq/pull/35774/commits/fcd44ab69032e981f640cfbb399fbda2123fbb92) for saving the user data.

ajeety4 commented on corehq/apps/integration/kyc/services.py:
```diff
@@ -129,17 +130,33 @@ def get_user_data_for_api(source, config):
         Returns a dictionary of user data for the API.
         ``source`` is a CommCareUser or a CommCareCase.
     """
+    # CommCareUser properties that could map to API fields
```
:1st_place_medal: 
> If a field is in a user's usercase, or custom user data, then use that,
otherwise fall back to a CommCareUser property. Only allow safe
properties.

This is great.

ajeety4 commented on corehq/apps/integration/kyc/forms.py:
```diff
@@ -45,7 +45,7 @@ class Meta:
         label=_('API Field to User Data Map'),
         required=True,
-        expected_type=list,
+        expected_type=dict,
```
This requires adding a migration file for it using  `python manage.py makemigrations integration`.

ajeety4 replied:
No worries

zandre-eng commented on corehq/apps/integration/kyc/services.py:
```diff
@@ -1,10 +1,141 @@
+    #         }
+    #     }
+
+    # TODO: Determine what thresholds we want
```
Will these thresholds be hard-coded, or will it be something a project can configure?

kaapstorm replied:
I think we can hard-code them in the beginning, and if we encounter a project that needs different thresholds, we can add them to KycConfig.

kaapstorm committed:
Add migration

kaapstorm committed:
Merge branch 'master' into nh/kyc/service

ajeety4 reviewed (approved):
I understand that we are yet to do a live test of the endpoint but I am good with the code changes.
I guess we can add a follow up ticket for the endpoint testing and some ToDo's.
